NTBS-133 manual results

jasiekmiko opened the pull request:
<!--- Provide a general summary of your changes in the Title above, including the related Jira ticket -->

## Description
Add manual test results data collection, including first instance of many-to-one data collection we are having for a notification.

*Note*: This is currently a WIP PR, with the following items still on their way:
- [x] visual tweaks, including: table order, background, button placement
- dynamic behaviour:
  - [x] js - filter list
  - [x] js - input validation
  - [x] js - disable test-carried-out input when there are items in it
- [x] deleting
- [x] audit (checking everything gets recorded as needed)
- [x] display on overivew

But it is a big one, so I'd definitely like to have it looked at at this stage and it *might* make sense to merge it in due to some of the more sweeping changes.

## Checklist:
- [ ] Automated tests are passing locally.
### Accessibility testing
- [x] Test functionality without javascript
- [ ] Test in small window (immitating small screen)
- [ ] Check the feature looks and works correctly in IE11
- [ ] Zoom page to 400% - content still visible?
- [ ] Test feature works with keyboard only operation
- [ ] Test with one screen reader (e.g. [NVDA](https://www.nvaccess.org/): see [getting started](https://webaim.org/articles/nvda/)) - logical flow, no unexpected content. 
- [ ] Passes automated checker (e.g. [WAVE](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh))

lbbethke commented on ntbs-integration-tests/NotificationPages/ClinicalDetailsPageTests.cs:
```diff
@@ -189,14 +188,13 @@ public async Task Post_RedirectsToNextPageAndSavesContent_IfModelValid()
             Assert.True(((IHtmlInputElement)reloadedDocument.GetElementById("NotificationSiteMap_PULMONARY_")).IsChecked);
             Assert.True(((IHtmlInputElement)reloadedDocument.GetElementById("NotificationSiteMap_OTHER_")).IsChecked);
             Assert.Equal("tissue", ((IHtmlInputElement)reloadedDocument.GetElementById("OtherSite_SiteDescription")).Value);
-            Assert.True(((IHtmlInputElement)reloadedDocument.GetElementById("ClinicalDetails_NoSampleTaken")).IsChecked);
```
Why was this line removed?

lbbethke commented on ntbs-integration-tests/NotificationPages/ManualTestResultEditPages.cs:
```diff
@@ -0,0 +1,167 @@
+using System.Collections.Generic;
```
Could we add "Test" to the filename?

lbbethke commented on ntbs-integration-tests/NotificationPages/ManualTestResultEditPages.cs:
```diff
@@ -0,0 +1,167 @@
+            var testsListDocument = await GetDocumentAsync(testsListPage);
+            var manualResults = testsListDocument.GetElementById("manual-results");
+
+            Assert.Contains("Smear", manualResults.TextContent);
```
Is it possible to grab the actual element that contains this by id/class, to be more specific?

jasiekmiko replied:
good shout!

lbbethke commented on ntbs-integration-tests/NotificationPages/ManualTestResultEditPages.cs:
```diff
@@ -0,0 +1,167 @@
+            var NotificationId = Utilities.DRAFT_ID;
+            var editUrl = GetCurrentPathForId(NotificationId) + TEST_ID;
+            var editPage = await Client.GetAsync(editUrl);
+            var editDocument = await GetDocumentAsync(editPage);
```
Could we assert that the field that is about to be changed has its original value here? It goes slightly against the arrange/act/assert but I think it would be good to add an additional assert step

lbbethke commented on ntbs-integration-tests/NotificationPages/ManualTestResultEditPages.cs:
```diff
@@ -0,0 +1,167 @@
+
+        [Fact]
+        public async Task PostEditWithInvalidResult_ReturnsErrors()
+        {
```
Could we rename this to "PostEditWithInvalidTestAndSampleTypeCombination_ReturnsErrors"? Might as well be more descriptive :) Might already not be the case in other tests (where I'm pretty sure I had fairly generic names) but looking through it as part of the review I feel like I'd prefer more precise naming

lbbethke commented on ntbs-service/Helpers/RouteHelper.cs:
```diff
@@ -33,5 +34,8 @@ public static class NotificationSubPaths
         public static string Delete => "Delete";
+
+        public static string EditManualTestResult(int? TestResultId) => $"Edit/ManualTestResult/{TestResultId}";
+        public static string AddManualTestResult => $"Edit/ManualTestResult";
```
It looks like overview page called TestResults, while the individual edit pages refers to ManualTestResult - Could we use the same name for both?

jasiekmiko replied:
The two are different: ManualTestResults are just one of the things that will end up on the TestResults page. Does that make it clearer?

lbbethke replied:
Ah ok, thanks for clarifying!

lbbethke commented on ntbs-service/Migrations/20191118112251_RemoveNoSampleFromClinicalDetails.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.DropColumn(
```
Ah that explains why you removed it from that test!

lbbethke commented on ntbs-service/Models/Enums/Result.cs:
```diff
@@ -0,0 +1,21 @@
+
+    public static class ResultEnumHelper
+    {
+        public static List<Result> GetAll()
```
I feel like we have a similar method elsewhere - is it possible to rewrite as a generic method taking type T : enum?

lbbethke replied:
Yes indeed we have one for status below!

jasiekmiko replied:
good catch! :)

lbbethke commented on ntbs-service/Models/ManualTestResult.cs:
```diff
@@ -0,0 +1,61 @@
+        public int ManualTestResultId { get; set; }
+        public int NotificationId { get; set; }
+        // We are not including a navigation property to Notification, otherwise it gets validated
+        // on every TryValidateModel(manualTestResult)
```
good to know!

lbbethke commented on ntbs-service/Models/ManualTestResult.cs:
```diff
@@ -0,0 +1,61 @@
+    public class ManualTestResult
+    {
+        // Even for values which are non-nullable in db, we make them a nullable runtime type so 
+        // the Required attribute can be applied properly, producing correct error messages
```
So what happens if you don't make them nullable - does it just give you a sql error rather than the error message? This will apply to my case too

jasiekmiko replied:
yes, i suppose it would. we never get there, because we validate the model before saving it to the db

lbbethke commented on ntbs-service/Pages/Notifications/Edit/ClinicalDetails.cshtml.cs:
```diff
@@ -93,22 +93,26 @@ private void SetupNotificationSiteMap(IEnumerable<NotificationSite> notification
+                (nameof(ClinicalDetails.DeathDate), FormattedDeathDate),
+                (nameof(ClinicalDetails.MDRTreatmentStartDate), FormattedMdrTreatmentDate)
+            }.ForEach(item => 
+                ValidationService.TrySetFormattedDate(ClinicalDetails, "ClinicalDetails", item.key, item.date)
```
Nice!

lbbethke commented on ntbs-service/Pages/Notifications/Edit/Table/ManualTestResult.cshtml.cs:
```diff
@@ -0,0 +1,132 @@
+            }
+        }
+
+        protected override async Task<IActionResult> PrepareAndDisplayPageAsync(bool isBeingSubmitted)
```
Could we move this to the top of the class? I think it's the first method defined for all the pages so makes it easier to compare

lbbethke commented on ntbs-service/Pages/Notifications/Edit/Table/ManualTestResult.cshtml.cs:
```diff
@@ -0,0 +1,132 @@
+        {
+            if (RowId != null)
+            {
+                TestResultForEdit = Notification.TestData.ManualTestResults
```
Can't we just fetch this directly from context.ManualTestResult.Find(RowId), rather than going via notification?

jasiekmiko replied:
Yeah could do - that actually feels better as we are then not requiring the Notification fetch to necessarily join onthe relevant info!

jasiekmiko replied:
ah actually, we would then have to manually check that the test belongs to the notification, which we get for free here - so i think i'll leave it as is.

lbbethke commented on ntbs-service/Pages/Notifications/Edit/Table/ManualTestResult.cshtml.cs:
```diff
@@ -0,0 +1,132 @@
+        private void SetDate()
+        {
+            // The required date will be marked as missing on the model, since we are setting it manually, rather than binding it
+            ModelState.Remove("TestResultForEdit.TestDate");
```
Do we have to do this elsewhere too where we set dates? Or is this case slightly different?

jasiekmiko replied:
I have checked other places for this and it didn't seem problematic. I think it is different in that we don't have require other dates, at best with `RequiredIf("ShouldValidateFull")`

lbbethke commented on ntbs-service/Pages/Notifications/Edit/Table/ManualTestResult.cshtml.cs:
```diff
@@ -0,0 +1,132 @@
+        public SelectList SampleTypes { get; set; }
+        public SelectList ResultOptions { get; set; }
+
+        [BindProperty(SupportsGet = true)]
```
Does this mean it's set from the url?

jasiekmiko replied:
yep :)

lbbethke commented on ntbs-service/Pages/Notifications/Edit/TestResults.cshtml.cs:
```diff
@@ -0,0 +1,69 @@
+
+        protected override async Task ValidateAndSave()
+        {
+            // Set the colleciton so it can be included in the validation
```
typo, collection

lbbethke commented on ntbs-service/Services/ValidationService.cs:
```diff
@@ -147,19 +150,19 @@ public bool IsValid(string key)
             if (formattedDate.TryConvertToDateTime(out DateTime? convertedDob))
             {
                 model.GetType().GetProperty(key).SetValue(model, convertedDob);
-                pageModel.TryValidateModel(model, modelTypeName);
```
Is that validation happening elsewhere now?

jasiekmiko replied:
Yes, all places that used this method called `TryValidateModel` themselves pretty much straight afterwards. I've checked that it stayed consistent between them

lbbethke replied:
Cool!

lbbethke commented on ntbs-ui-tests/Features/CreateNotifications.feature:
```diff
@@ -41,7 +41,21 @@ Feature: Notification creation
         And I enter 1 into 'FormattedDiagnosisDate_Month'
```
Are you on a branch before my autocomplete changes? As that broke somethign (because autocomplete is not a normal select input) - will fix as part of my ticket

jasiekmiko replied:
umm, most likely - yes :P 
is there anything for me to do here?

lbbethke replied:
No I'll fix them, have already started looking into how!

jasiekmiko commented:
Thanks for the thorough review @lbbethke !
I've just added deleting, too.
I think once I add some small visual tweaks we discussed with Nancy, and I've addressed your comments above, this will be good enough to merge and should unblock you, so i'll make that my priority tomorrow.

lbbethke commented on ntbs-service/Pages/Notifications/Edit/Table/ManualTestResult.cshtml.cs:
```diff
@@ -47,27 +47,49 @@ protected override async Task ValidateAndSave()
+            Notification = await GetNotificationAsync(NotificationId);
+            if (!(await AuthorizationService.CanEdit(User, Notification)))
+            {
+                return ForbiddenResult();
```
Oh this makes me think that we need to add a 403 page after all which is identical to the access denied page - because our new error handler will be looking for this when ForbiddenResult is thrown!

jasiekmiko replied:
yes, i think it'd be good to still have it available for situation when we do want to throw it! :)

lbbethke commented on ntbs-service/Services/TestResultsRepository.cs:
```diff
@@ -26,9 +29,17 @@ public async Task AddTestResultAsync(ManualTestResult testResult)
-        public async Task UpdateTestResultAsync(ManualTestResult testResult)
+        public async Task UpdateTestResultAsync(Notification Notification, ManualTestResult testResult)
         {
-            _context.Attach(testResult);
```
Why did you change this? Did it not update properly, or was it an auditing thing? I feel like functionally both should be equivalent

jasiekmiko replied:
Because I already had the context-attached model, and I was trying to attach another instance with the same id. The way it does it now follows what we do for notification - we pass in the context-attached model, and a model with the new data we want.

I've not spotted it before because my add/update logic was backwards 🤦‍♂️

lbbethke replied:
I see, makes sense!

jasiekmiko commented:
@conorsheehansw passing this back to you now. I've updated the PR description with what I believe to be left on it!

conorsheehansw commented on ntbs-service/Pages/Notifications/Edit/Children/ManualTestResult.cshtml:
```diff
@@ -13,121 +13,127 @@
         <form method="post" autocomplete="off">
             <nhs-grid-row>
-                <nhs-grid-column grid-column-width="OneHalf">
+                <nhs-grid-column grid-column-width="Full">
```
Semantically slightly iffy use of columns, but I think it has a better result on render - don't know if it was mandated with Nancy to push these onto 2 lines - I just don't think the date plays well with other inputs on the same line.
I tried thirds for the other inputs, and that looked a bit iffy too.

jasiekmiko replied:
I think the semantics are actually fine, in fact, i would consider using a single row with all columns as you have them.
as for the changed layout - do verify with Nancy, but LGTM!

conorsheehansw commented:
@jasiekmiko 
Pretty much the only thing I managed to move to what I consider complete (you might disagree with this) is:
- [X]  js - disable test-carried-out input when there are items in it

I interpreted this slightly different to the written: 

- Disabled server-side 'tests carried out' input if any manual test results are populated.
- Added progressive enhancement for hiding the manual test results content if false is the selected value.

Encountered a bit of an issue with date validation (and will be an issue for the filtering lists too) that I sunk a bit of time into but didn't come up with anything I was happy with. In the page route because we have an optional route parameter (`rowId?`) before `handler?` when we're in add mode the handler doesn't get picked up...

Did a bit of googling and tried a couple of the suggestions but couldn't come up with anything concrete without sacrificing the current route...

jasiekmiko commented on ntbs-integration-tests/NotificationPages/ManualTestResultEditPagesTests.cs:
```diff
@@ -54,8 +57,8 @@ public static IList<Notification> GetSeedingNotifications()
+            const int notificationId = Utilities.DRAFT_ID;
+            var url = GetCurrentPathForId(notificationId);
             var initialPage = await Client.GetAsync(url);
             var initialDocument = await GetDocumentAsync(initialPage);
```
You might want to update to the `GetDocumentForUrl` helper from Laura here (and rest of the file)

jasiekmiko commented on ntbs-integration-tests/NotificationPages/ManualTestResultEditPagesTests.cs:
```diff
@@ -210,10 +213,58 @@ public async Task PostDelete_ReturnsSuccessAndRemovesResult()
+            const int manualTestTypeId = (int)ManualTestTypeId.Histology;
+            const string handlerPath = "FilteredSampleTypesForManualTestType";
+            var endpointPath = $"{NotificationSubPath}{(isEdit ? "0/" : "")}{handlerPath}";
+            var endpointUrl = GetPathForId($"{endpointPath}?value={manualTestTypeId.ToString()}", 0);
```
I think it's a worry that we're constructing a query string ourselves here, whilst `GetPathForId` is capable of it too. It's not a codepath we are hitting currently, but in theory we could end up with multiple `?` signs in the url.
Should we let `GetPathForId` accept an optional dictionary of query string values instead, and passing it through to a single place that does url construction?

jasiekmiko commented on ntbs-service/Pages/Notifications/Edit/Children/ManualTestResult.cshtml:
```diff
@@ -1,153 +1,8 @@
 @page "/Notifications/{NotificationId:int}/Edit/ManualTestResult/{RowId:int?}/{handler?}"
```
Can `RowId` become non-null now?

jasiekmiko commented on ntbs-service/Pages/Notifications/Edit/Children/NewManualTestResult.cshtml:
```diff
@@ -0,0 +1,8 @@
+@page "/Notifications/{NotificationId:int}/Edit/ManualTestResult/{handler?}"
```
Since we've arrived at splitting these up anyways, do we want to have `/Edit/ManualTestResult/New` as the page url here?

lbbethke replied:
Ahh see my question in the tests - is this what fixes the dynamic validation issues for new test results? I should probably merge in the most recent changes into my social context venue branch...

jasiekmiko commented on ntbs-service/Pages/Notifications/Edit/Children/NewManualTestResult.cshtml:
```diff
@@ -0,0 +1,8 @@
+@model ManualTestResultPage
+
+@{
+    Layout = "../../../Shared/_NotificationLayout.cshtml";
```
We can give it a nice and custom page name now too!

jasiekmiko commented on ntbs-service/wwwroot/css/site.scss:
```diff
@@ -169,3 +169,16 @@ body {
+        margin-top: 18px;
+        margin-bottom: 40px;
+    }
+}
```
Do we want them shrunk even more on smaller screens?
If not, I **think** we can combine the selectors to something along the lines of:
```
.nhsuk-table-responsive {
  &, @media (min-width: 40.0625em) {
        margin-top: 18px;
        margin-bottom: 40px;
    }
} 
```
... maybe?

jasiekmiko commented on ntbs-service/wwwroot/source/Components/FilteredDropdown.ts:
```diff
@@ -0,0 +1,122 @@
+            }
+
+            axios.request(requestConfig)
+                .then((response: any) => {
```
Is it worth giving the response a type?

conorsheehansw replied:
Think we could only specify it down to `{ data: any }` - so not sure it's worth it.
Currently data either contains `{Hospital: [], CaseManagers:[]}` or `{SampleTypes:[]}`

jasiekmiko commented on ntbs-service/wwwroot/source/Components/FilteredDropdown.ts:
```diff
@@ -0,0 +1,122 @@
+    group: string
+};
+
+const FilteredDropdown = Vue.extend({
```
I feel like this component could use a short comment above to explain how to use it when wiring it up, what do you think?

lbbethke commented on ntbs-integration-tests/NotificationPages/ManualTestResultEditPagesTests.cs:
```diff
@@ -210,10 +213,58 @@ public async Task PostDelete_ReturnsSuccessAndRemovesResult()
+            };
+            const string handlerPath = "ValidateTestResultForEditDate";
+            var queryString = string.Join("&", formData.Select(kvp => $"{kvp.Key}={kvp.Value}"));
+            var endpointPath = $"{NotificationSubPath}{(isEdit ? "0/" : "")}{handlerPath}";
```
Why is the Id 0 for the edit case, shouldn't it be the actual id of the manual test type? Also, does this path actually resolve for the case where no id is provided?
I'm working on top off this branch 3 commits behind, and in my checkout the dynamic validation for a NEW manual test result (i.e. where there is no id in the path) does not work as the path doesn't get resolved - if it works for you, do you know what fixed it? I'm doing the same thing for social context venues.

conorsheehansw replied:
You've already spotted the split between edit/add. 
Id = 0, as we don't actually care what RowId is, we're just hitting the dynamic endpoints accessed normally through components.

lbbethke replied:
Yep, I realised the actual value of the Id didn't matter after I wrote the comment
